pkg/receiptsmigration/migration.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2019 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receiptsmigration
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/krew/pkg/constants"
+	"sigs.k8s.io/krew/pkg/environment"
+	"sigs.k8s.io/krew/pkg/index"
+	"sigs.k8s.io/krew/pkg/receiptsmigration/oldenvironment"
+)
+
+const (
+	krewPluginName = "krew"
+)
+
+// Done checks if the krew installation requires a migration.
+// It considers a migration necessary when plugins are installed, but no receipts are present.
+func Done(newPaths environment.Paths) (bool, error) {
+	receipts, err := ioutil.ReadDir(newPaths.InstallReceiptPath())
+	if err != nil {
+		return false, err
+	}
+	store, err := ioutil.ReadDir(newPaths.InstallPath())
+	if err != nil {
+		return false, err
+	}
+
+	var hasInstalledPlugins bool
+	for _, entry := range store {
+		if entry.IsDir() {
+			hasInstalledPlugins = true
+			break
+		}
+	}
+
+	hasNoReceipts := len(receipts) == 0
+
+	return !(hasInstalledPlugins && hasNoReceipts), nil
+}
+
+// Migrate searches for installed plugins, removes each plugin and reinstalls afterwards.
+// Once started, it keeps going even if there are errors.
+func Migrate(newPaths environment.Paths) error {
+	isMigrated, err := Done(newPaths)
+	if err != nil {
+		return err
+	}
+	if isMigrated {
+		glog.Infoln("Already migrated")
+		return nil
+	}
+
+	oldPaths := oldenvironment.MustGetKrewPaths()
+	installed, err := getPluginsToReinstall(oldPaths, newPaths)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build list of plugins")
+	}
+
+	glog.Infoln("These plugins will be reinstalled: ", installed)
+	if err := os.MkdirAll(newPaths.InstallReceiptPath(), 0755); err != nil {
+		return errors.Wrapf(err, "failed to create directory %q", newPaths.InstallReceiptPath())
+	}
+
+	// krew must be skipped by the normal migration logic
+	if err := copyKrewManifest(newPaths.IndexPluginsPath(), newPaths.InstallReceiptPath()); err != nil {
+		return errors.Wrapf(err, "failed to copy krew manifest")
+	}
+
+	for _, plugin := range installed {
+		if consistent := isInstallationConsistent(oldPaths.BinPath(), plugin); !consistent {
```

Comment:
why would this case happen?

Reply:
I wanted to guard against the case where a user has added other files in `store` or `bin`. But probably the check that the plugin is available in the index already covers that well enough. I'll remove the check.